orttraining/orttraining/training_api/checkpoint.h

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(ENABLE_TRAINING) && defined(ENABLE_TRAINING_ON_DEVICE)
+
+#pragma once
+#include "core/platform/path_lib.h"
+#include "core/platform/env.h"
+#include "orttraining/core/framework/checkpointing.h"
+#include "orttraining/training_api/interfaces.h"
+#include <type_traits>
+#include "onnx/defs/tensor_proto_util.h"
+
+namespace onnxruntime {
+namespace training {
+
+// TODO: Rename to api after all major classes implemented.
+namespace api_test {
+
+/**
+ * @brief Base class for user defined checkpoint property.
+ */
+struct CheckpointProperty {
+ public:
+  CheckpointProperty(const std::string& prop_name)
+      : prop_name_(prop_name) {
+  }
+
+  virtual ~CheckpointProperty() {}
+  virtual ONNX_NAMESPACE::TensorProto ToTensorProto() = 0;
+
+  std::string GetName() const {
+    return prop_name_;
+  }
+
+ protected:
+  std::string prop_name_;
+};
+
+/**
+ * @brief User defined checkpoint property.
+ * Only support int64_t, std::string and float data types.
+ */
+template <typename T>
+struct TypedCheckpointProperty : public CheckpointProperty {
+ public:
+  TypedCheckpointProperty(const std::string& prop_name, const T& prop_value)
+      : CheckpointProperty(prop_name), prop_value_(prop_value) {
+  }
+
+  ONNX_NAMESPACE::TensorProto ToTensorProto() override {
+    auto t_proto = ONNX_NAMESPACE::ToTensor<T>(prop_value_);
+
+    // We did not add dims assuming currently we only support scalars.
+    return t_proto;
+  }
+
+  T GetData() const {
+    return prop_value_;
+  }
+
+ private:
+  T prop_value_;
+};
+
+/**
+ * @brief A data class representing traing states.
+ * Including:
+ *     > Parameter states.
+ *     > Optimizer states.
+ *     > User defined training properties, for example 'epoch', 'best_score'.
+ */
+struct CheckpointStates {
+ public:
+  ModuleCheckpointStates module_checkpoint_states;
+  OptimizerCheckpointStates optimizer_checkpoint_states;
+  std::unordered_map<std::string, std::shared_ptr<CheckpointProperty>> named_properties;
+};
+
+PathString CreateFolderIfNotExists(const PathString& path, const std::string& folder_name);
+
+Status CreateOrtValuesFromTensorProtos(
+    const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+    NameMLValMap& name_to_ort_value);
+
+/**
+ * @brief A single entry for checkpoint utilities.
+ *
+ * A checkpoint is a directory of files:
+ * checkpoint/
+ *   parameters/
+ *       tensors.pbseq - tensor protobuf messages
+ *       tensors.bin - tensor binary data
+ *   optimizers/
+ *       group_0/
```

Comment:
A group of parameters will share same optimizer hyperparameters including learning rate. Currently we don't have user requirement to support more than one group, but my thinking is we'd better make the ckpt schema extensible to support it  later.

Reply:
Is it possible to embed this information inside the metadata instead of structuring the checkpoint around it? It might become difficult to maintain this structure in the future when we build our more features to meet requirements (example zero or distributed training and so on).
Keeping this information in a metadata file might be easier to consume and maintain.

Reply:
* A checkpoint is a directory of files:
 * checkpoint/
 *   paramtrain_tensors.pbseq - trainable parameter tensor protobuf messages
 *   paramfrozen_tensors.pbseq - non_trainable parameter tensor protobuf messages
 *   optim_group0_momentum0_tensors.pbseq - optimizer momentum state tensor protobuf messages
 *   optim_group0_momentum1_tensors.pbseq - optimizer momentum state tensor protobuf messages
 *   optim_group0_properties.pbseq - group-wise optimizer property tensor protobuf messages
 *   custom_properties.pbseq - custom property protobuf messages

Did not use the folder hierarchy anymore, instead, the file name contains the states name, group name, etc. 
Let's refine into a meta data based approach once we need it.